external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp

```diff
@@ -916,6 +919,46 @@ struct BFOpLowering : ConvertToLLVMPattern {
     return success();
   }
 };
+
+struct WarpSwizzleOpLowering : ConvertToLLVMPattern {
+  explicit WarpSwizzleOpLowering(MLIRContext *context,
+                                 LLVMTypeConverter &typeConverter)
+      : ConvertToLLVMPattern(gpu::WarpSwizzleOp::getOperationName(), context,
+                             typeConverter) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto swizzleOp = cast<gpu::WarpSwizzleOp>(op);
+    auto loc = swizzleOp.getLoc();
+    auto adaptor = gpu::WarpSwizzleOpAdaptor(operands, op->getAttrDictionary());
+
+    auto mlirI32Type = rewriter.getI32Type();
+    auto llvmI32Type = typeConverter->convertType(mlirI32Type);
+    auto llvmI1Type = typeConverter->convertType(rewriter.getI1Type());
+
+    int32_t permConst = 0;
+    const ArrayRef<mlir::Attribute> selector = adaptor.selector().getValue();
+    for (auto v = selector.rbegin(); v != selector.rend(); ++v) {
+      permConst = (permConst << 2) |
+                  v->cast<mlir::IntegerAttr>().getValue().getZExtValue();
+    }
+
+    auto dppCtrlConstImm = rewriter.create<LLVM::ConstantOp>(
+        loc, llvmI32Type, rewriter.getI32IntegerAttr(permConst));
+
+    auto noMaskConstImm = rewriter.create<LLVM::ConstantOp>(
+        loc, llvmI32Type, rewriter.getI32IntegerAttr(0b1111));
```

Comment:
this magic constant value probably deserves a comment.

Reply:
Fixed